include/swift/AST/AutoDiff.h

```diff
@@ -101,145 +101,41 @@ class ParsedAutoDiffParameter {
-  ///
-  ///   functionType = (A, B, C) -> R
-  ///   if "A" and "C" are in the set,
-  ///   ==> pushes {A, C} to `paramTypes`.
-  ///
-  ///   functionType = (A, B) -> (C, D) -> R
-  ///   if "A", "C", and "D" are in the set,
-  ///   ==> pushes {A, C, D} to `paramTypes` if `reverseCurryLevels` is false,
-  ///    or pushes {C, D, A} otherwise.
-  ///
-  ///   functionType = (Self) -> (A, B, C) -> R
-  ///   if "Self" and "C" are in the set,
-  ///   ==> pushes {Self, C} to `paramTypes` if `reverseCurryLevels` is false,
-  ///    or pushes {C, Self} otherwise.
-  ///
-  void getSubsetParameterTypes(AnyFunctionType *functionType,
-                               SmallVectorImpl<Type> &paramTypes,
-                               bool reverseCurryLevels = false) const;
-
-  /// Returns a bitvector for the SILFunction parameters corresponding to the
-  /// parameters in this set. In particular, this explodes tuples. For example,
-  ///
-  ///   functionType = (A, B, C) -> R
-  ///   if "A" and "C" are in the set,
-  ///   ==> returns 101
-  ///   (because the lowered SIL type is (A, B, C) -> R)
-  ///
-  ///   functionType = (Self) -> (A, B, C) -> R
-  ///   if "Self" and "C" are in the set,
-  ///   ==> returns 0011
-  ///   (because the lowered SIL type is (A, B, C, Self) -> R)
-  ///
-  ///   functionType = (A, (B, C), D) -> R
-  ///   if "A" and "(B, C)" are in the set,
-  ///   ==> returns 1110
-  ///   (because the lowered SIL type is (A, B, C, D) -> R)
-  ///
-  AutoDiffIndexSubset *getLowered(ASTContext &ctx,
-                                  AnyFunctionType *functionType) const;
-
-  void Profile(llvm::FoldingSetNodeID &ID) const {
-    ID.AddInteger(parameters.size());
-    for (unsigned setBit : parameters.set_bits())
-      ID.AddInteger(setBit);
-  }
-};
+class AnyFunctionType;
+class AutoDiffIndexSubset;
+class Type;
+class SILModule;
+enum class SILLinkage : uint8_t;
 
-/// Builder for `AutoDiffParameterIndices`.
-class AutoDiffParameterIndicesBuilder {
+/// Builder for `AutoDiffIndexSubset`.
+class AutoDiffIndexSubsetBuilder {
   llvm::SmallBitVector parameters;
-
+  
 public:
-  /// Creates a `AutoDiffParameterIndicesBuilder` for the given function type.
-  AutoDiffParameterIndicesBuilder(AnyFunctionType *functionType);
-
-  /// Builds the `AutoDiffParameterIndices`, returning a pointer to an existing
+  /// Creates a `AutoDiffIndexSubsetBuilder` for the given function type.
+  AutoDiffIndexSubsetBuilder(AnyFunctionType *functionType);
+  
+  /// Builds the `AutoDiffIndexSubset`, returning a pointer to an existing
   /// one if it has already been allocated in the `ASTContext`.
-  AutoDiffParameterIndices *build(ASTContext &C) const;
-
-  /// Sets the parameter at `parameterIndex`. See
-  /// `AutoDiffParameterIndices::parameters` for documentation about the order.
+  AutoDiffIndexSubset *build(ASTContext &C) const;
+  
+  /// Sets the parameter at `parameterIndex`.
   void setParameter(unsigned parameterIndex);
-
+  
```

Comment:
Remove trailing whitespaces.

Reply:
Done in https://github.com/apple/swift/pull/27514/commits/37d1a911cd33f0a3bc45ce1fe0499727ed04c60a.